From dc0d9d915d27f1c55b7385e64ce81b607e297f21 Mon Sep 17 00:00:00 2001
From: Josef Schlehofer <pepe.schlehofer@gmail.com>
Date: Sun, 9 Jan 2022 20:41:21 +0100
Subject: [PATCH] mvebu: Add support for Turris MOX

This patch adds support for CZ.NIC Turris MOX (the modular router).

Specifications of module A (the base one):

- SoC: Marvell Armada 3720, ARMv8, dual-core 1 GHz
- RAM: 512 MB or 1 GB DDR3
- Flash: 8MB SPI NOR (w25q64dw)
- SDIO bus for Wi-Fi addon
- 1x USB 3.0 port
- 1x 1Gbps RJ45  port
- 1x RTC clock (Microchip MPC7940x)
- microSD card slot
- 34-pin GPIO CONNECTOR
- 8-pin PoE bracker
- 65-pin MOXTET to connect another modules
module B - miniPCIe slot
module C - 4x Gigabit Ethernet ports
module D - SFP module
module E - 8x Gigabit Ethernet ports
module G - miniPCIe slot
module F - 4x USB 3.0 ports

Signed-off-by: Josef Schlehofer <pepe.schlehofer@gmail.com>
---
 package/boot/uboot-tools/files/mvebu                |  3 +++
 .../mvebu/cortexa53/base-files/etc/board.d/02_network  |  3 +++
 target/linux/mvebu/cortexa53/config-5.10               |  5 +++++
 target/linux/mvebu/image/cortexa53.mk                  | 10 ++++++++++
 4 files changed, 21 insertions(+)

diff --git a/package/boot/uboot-tools/files/mvebu b/package/boot/uboot-envtools/files/mvebu
index cffd396ae1..8c05c91cf7 100644
--- a/package/boot/uboot-tools/files/mvebu
+++ b/package/boot/uboot-tools/files/mvebu
@@ -26,6 +26,9 @@ cznic,turris-omnia)
 		ubootenv_add_uci_config "/dev/mtd0" "0xf0000" "0x10000" "0x10000"
 	fi
 	;;
+cznic,turris-mox)
+	ubootenv_add_uci_config "/dev/mtd2" "0x0" "0x00010000"
+	;;
 glinet,gl-mv1000)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x8000" "0x8000" "1"
 	;;
diff --git a/target/linux/mvebu/cortexa53/base-files/etc/board.d/02_network b/target/linux/mvebu/cortexa53/base-files/etc/board.d/02_network
index 7da9de58c7..8b54a3bcc7 100644
--- a/target/linux/mvebu/cortexa53/base-files/etc/board.d/02_network
+++ b/target/linux/mvebu/cortexa53/base-files/etc/board.d/02_network
@@ -10,6 +10,9 @@ board_config_update
 board=$(board_name)
 
 case "$board" in
+cznic,turris-mox)
+	ucidef_set_interface_lan_wan "$(cd /sys/class/net/; echo lan*)" "eth0"
+	;;
 glinet,gl-mv1000|\
 globalscale,espressobin|\
 globalscale,espressobin-emmc|\
diff --git a/target/linux/mvebu/cortexa53/config-5.10 b/target/linux/mvebu/cortexa53/config-5.10
index a47d66c669..7977ab03f9 100644
--- a/target/linux/mvebu/cortexa53/config-5.10
+++ b/target/linux/mvebu/cortexa53/config-5.10
@@ -40,6 +40,7 @@ CONFIG_GENERIC_BUG_RELATIVE_POINTERS=y
 CONFIG_GENERIC_CPU_VULNERABILITIES=y
 CONFIG_GENERIC_CSUM=y
 CONFIG_GENERIC_PINCONF=y
+CONFIG_GPIO_MOXTET=y
 CONFIG_HOLES_IN_ZONE=y
 CONFIG_ILLEGAL_POINTER_VALUE=0xdead000000000000
 CONFIG_MAILBOX=y
@@ -47,6 +48,10 @@ CONFIG_MAILBOX=y
 CONFIG_MFD_SYSCON=y
 CONFIG_MMC_SDHCI_XENON=y
 CONFIG_MODULES_USE_ELF_RELA=y
+CONFIG_MOXTET=y
+CONFIG_MTD_M25P80=y
+CONFIG_MTD_SPI_NOR=y
+CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y
 CONFIG_MVEBU_GICP=y
 CONFIG_MVEBU_ICU=y
 CONFIG_MVEBU_ODMI=y
diff --git a/target/linux/mvebu/image/cortexa53.mk b/target/linux/mvebu/image/cortexa53.mk
index ca3a53f992..fa22de063c 100644
--- a/target/linux/mvebu/image/cortexa53.mk
+++ b/target/linux/mvebu/image/cortexa53.mk
@@ -1,3 +1,13 @@
+define Device/cznic_turris-mox
+  $(call Device/Default-arm64)
+  DEVICE_VENDOR := CZ.NIC
+  DEVICE_MODEL := Turris MOX
+  DEVICE_PACKAGES := kmod-usb2 kmod-gpio-button-hotplug kmod-rtc-ds1307 kmod-i2c-pxa
+  SOC := armada-3720
+  BOOT_SCRIPT := turris-mox
+endef
+TARGET_DEVICES += cznic_turris-mox
+
 define Device/glinet_gl-mv1000
   $(call Device/Default-arm64)
   DEVICE_VENDOR := GL.iNet
-- 
2.32.0

